hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantUserCreateRequest.java

```diff
@@ -0,0 +1,285 @@
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.helpers.S3SecretValue;
+import org.apache.hadoop.ozone.om.multitenant.OzoneMultiTenantPrincipal;
+import org.apache.hadoop.ozone.om.ratis.utils.OzoneManagerDoubleBufferHelper;
+import org.apache.hadoop.ozone.om.request.util.OmResponseUtil;
+import org.apache.hadoop.ozone.om.request.volume.OMVolumeRequest;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.om.response.s3.tenant.OMTenantCreateResponse;
+import org.apache.hadoop.ozone.om.response.s3.tenant.OMTenantUserCreateResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.CreateTenantUserRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.CreateTenantUserResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.S3Secret;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.UpdateGetS3SecretRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.S3_SECRET_LOCK;
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.VOLUME_LOCK;
+
+/*
+  Ratis execution flow for OMTenantUserCreate
+
+- Client (UserCreateHandler , etc.)
+  - Check username validity: ensure no invalid characters
+  - Send request to server
+- OMTenantUserCreateRequest
+  - preExecute (perform checks and init)
+    - Check username validity (again), check $
+      - If username is invalid, throw exception to client; else continue
+    - Generate S3 secret for the new user
+  - validateAndUpdateCache (update DB)
+    - Permission check (checkACL need to check access key now)
+    - Grab VOLUME_LOCK write lock
+    - Check user existence
+      - If user doesn't exist, throw exception to client; else continue
+    - Check tenant existence
+      - If tenant doesn't exist, throw exception to client; else continue
+    - Grab S3_SECRET_LOCK write lock
+    - S3SecretTable: Flush generated S3 secret
+      - Key: TENANTNAME$USERNAME (equivalent to kerberosID)
+      - Value: <GENERATED_SECRET>
+    - Release S3_SECRET_LOCK write lock
+    - tenantUserTable: New entry
+      - Key: Tenant user name. e.g. finance$bob, s3v$alice
+      - Value: Tenant name. e.g. finance
+    - tenantGroupTable: Add this new user to the default tenant group.
+      - Key: finance$bob
+      - Value: finance-users
+    - tenantRoleTable: TBD. NoOp for prototype.
+    - Release VOLUME_LOCK write lock
+ */
+
+/**
+ * Handles OMTenantUserCreate request.
+ */
+public class OMTenantUserCreateRequest extends OMVolumeRequest {
```

Comment:
Is there a plan to provide a way get the s3 creds of a user from command line post creation? Something like an extension to the **ozone s3 getsecret** command.

Reply:
Yes we'd better have one as of right now the only way is retrieve the creds after tenant user creation is to dump the DB.. maybe allow it for admin at least

Reply:
yup